Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2202,12 +2202,111 @@ def visit_FuncDefNode(self, node):
         self._super_visit_ClassDefNode(node)
         self.qualified_name = orig_qualified_name
         return node
 
+class MangleDunderNamesTransform(CythonTransform):
+    """__names appearing inside classes are mangled to their _classname__names form
+
+    This should be called after the pure Python @cython directives have all been
+    applied"""
+    def __init__(self, *args, **kwargs):
+        super(MangleDunderNamesTransform, self).__init__(*args, **kwargs)
+        self.last_class_name = None
+        self.dont_mangle_pyx = False
+
+    def _visit_class_common(self, node, name):
+        # The class name (target), list of bases and metaclass name should be
+        # processed without being mangled by the class itself.
+        # Note that for Python 2 the exclusion of the metaclass is incorrect
+        # since it is defined inside the class as __metaclass__
+        exclude_at_first = ['bases', 'metaclass', 'target']
+        attrs = [ attr for attr in node.child_attrs if attr not in exclude_at_first ]
+
+        prev = self.last_class_name
+        self.last_class_name = name
+        self.visitchildren(node, attrs=attrs)
+        self.last_class_name = prev
+
+        self.visitchildren(node, attrs=exclude_at_first)
+        return node
+
+    def visit_ModuleNode(self, node):
+        dont_mangle_pyx = self.dont_mangle_pyx
+        # classes defined in utility code are treated slightly and attributes
+        # starting with __pyx_ aren't mangled (otherwise memoryviews break)
+        self.dont_mangle_pyx = isinstance(node.scope, NonManglingModuleScope)
+        self.visitchildren(node)
+        self.dont_mangle_pyx = dont_mangle_pyx
+        return node
+
+    def visit_PyClassDefNode(self, node):
+        return self._visit_class_common(node, node.name)
+
+    def visit_CClassDefNode(self, node):
+        # although neither metaclass or target exist for CClassDefNode this
+        # does not prevent them being treated in the same way
+        if node.visibility == "extern": # "public" too?
+            # it's being imported from elsewhere so names are already set
+            return node
+        else:
+            return self._visit_class_common(node, node.class_name)
+
+    def visit_NameNode(self, node):
+        node.name = self.mangle_special_name(node.name)
+        self.visitchildren(node)
+        return node
+
+    def visit_CNameDeclaratorNode(self, node):
+        return self.visit_NameNode(node)
+
+    def visit_DefNode(self, node):
+        return self.visit_NameNode(node)
+
+    def visit_AttributeNode(self, node):
+        node.attribute = self.mangle_special_name(node.attribute)
+        self.visitchildren(node)
+        return node
+
+    def visit_CompilerDirectivesNode(self, node):
+        # It probably isn't desirable to wait until ForwardDeclareTypes to
+        # mangle the names (the earlier the better as far as possible).
+        # Therefore, attempt to change the relevant compiler directives
```

Comment:
I would prefer having one place to evaluate these directives, and not pre-process them in a different place. They should first get mapped to normal declarations, and then those should be processed in the same way as the `cdef` and Python declarations.

Reply:
I'm struggling to see how to get rid of this. The `locals` directive seems to be evaluated in `AnalyseDeclarationsTransform` (in contrast to what I've put in the comment...). By that time the names are already being put into scope entries tables, so it's too late for this transform to deal with them.

It's either what I've done here, or a big change to how the locals directive is processed.

Reply:
I've left this as is for the moment because I couldn't find a good way of getting rid of it. I think the only sensible way to do it would be to move the name mangling back into Symtab (which probably isn't hugely difficult, if necessary)

Reply:
Well … name mangling has always been done in Symtab. If this can be resolved in there as well, I think that's desirable. It's a thing that's highly related to scopes, and that's what Symtab is all about.